In [1]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [2]:
# Load the given datasets
ckd_categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")
ckd_numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")
# Print the data
print(ckd_categorical)

print(ckd_numerical)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: 
This data is different factors related to blood, kidneys, and related systems that may affect chronic kidney disease.
There are features for age, blood pressure, specific gravity, albumin, sugar, red blood cells, pus cell, pus cell clumps, bacteria, blood glucose random, blood urea, serum creatinine, sodium, potassium, hemoglobin, packed cell volume, white blood cell count, red blood cell count, hypertension, diabetes mellitus, coronary artery disease, appetite, pedal edema, and [ckd or not ckd].
The features in the numerical dataset are numerical, and the features in the categorical dataset are categorical. Two features in the categorical set (al and su) are represented using numbers, but are still categorical in nature.
Categorical Examples: Target, pcc, ba
Numerical Examples: age, bp, wbcc

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
#print(ckd_categorical)

visited = []
duplicate_rows = []
dup_count = 0
counter = 0
for id in ckd_numerical.loc[:,'unique_id']:
    if id not in visited:
        visited.append(id)
    else:
        dup_count += 1
        duplicate_rows.append(counter)
    counter += 1

print(dup_count,"duplicate rows")

# Drop these duplicate rows
duplicate_rows.sort(reverse=True)
for row in duplicate_rows:
    ckd_numerical.drop(row, inplace = True)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
visited = []
duplicate_rows = []
dup_count = 0
counter = 0
for id in ckd_categorical.loc[:,"unique_id"]:
    if id not in visited:
        visited.append(id)
    else:
        dup_count += 1
        duplicate_rows.append(counter)
        print("duplicate row at:", counter)
    counter += 1

print(dup_count,"duplicate rows")

duplicate_rows.sort(reverse=True)
for row in duplicate_rows:
    ckd_categorical.drop(row, inplace = True)

#ckd_categorical

30 duplicate rows
duplicate row at: 231
duplicate row at: 401
duplicate row at: 402
duplicate row at: 403
duplicate row at: 404
duplicate row at: 405
duplicate row at: 406
duplicate row at: 407
duplicate row at: 408
duplicate row at: 409
duplicate row at: 410
duplicate row at: 411
duplicate row at: 412
duplicate row at: 413
duplicate row at: 414
duplicate row at: 415
duplicate row at: 416
duplicate row at: 417
duplicate row at: 418
duplicate row at: 419
duplicate row at: 420
duplicate row at: 421
duplicate row at: 422
duplicate row at: 423
duplicate row at: 424
duplicate row at: 425
duplicate row at: 426
duplicate row at: 427
duplicate row at: 428
duplicate row at: 429
30 duplicate rows


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [4]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
ckd = ckd_numerical
ckd = ckd.merge(ckd_categorical, how='outer', on = "unique_id")

#Print the combined dataset
print(ckd)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [5]:
# Calculate the percentage of rows that contain atleast one missing value
num_rows = ckd.shape[0]
#print(num_rows)
#print(ckd.isnull())
empty_rows = ckd.isnull().any(axis = 1)
#print(empty_rows)
missing_count = 0
for row in empty_rows:
    if row == True:
        missing_count += 1
#print(missing_count)
percent_na = (missing_count/num_rows)*100




# Print %
print(percent_na)
# Drop these rows from the dataset
ckd.dropna(inplace = True)

# Print the Dataset
print(ckd)

61.75000000000001
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc        

## Part 3.4: Sort the dataset according to the Labels

In [6]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
ckd.sort_values(by="Target", inplace = True)
ckd.reset_index(inplace = True)

# Print the dataset
print(ckd)

     index  unique_id   age    bp    bgr     bu    sc    sod  pot  hemo  ...  \
0        0     102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  ...   
1      103     337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  ...   
2      105     343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  ...   
3      108     349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  ...   
4      110     353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  ...   
..     ...        ...   ...   ...    ...    ...   ...    ...  ...   ...  ...   
148    171     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  ...   
149     43     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  ...   
150    169     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  ...   
151     54     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  ...   
152    396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  ...   

           pc         pcc          ba  

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [7]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
ckd = pd.get_dummies(ckd, categorical_columns)

# Print the dataset
print(ckd)

     index  unique_id   age    bp    bgr     bu    sc    sod  pot  hemo  ...  \
0        0     102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  ...   
1      103     337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  ...   
2      105     343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  ...   
3      108     349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  ...   
4      110     353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  ...   
..     ...        ...   ...   ...    ...    ...   ...    ...  ...   ...  ...   
148    171     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  ...   
149     43     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  ...   
150    169     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  ...   
151     54     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  ...   
152    396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  ...   

     cad_no  cad_yes  appet_good  appet

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [8]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for col in numerical_columns:
    
    stdev = ckd.loc[:, col].std()
    mean = ckd.loc[:, col].mean()
    #print(stdev)
    outlier_rows = []
    counter = 0
    for val in ckd.loc[:,col]:
        if abs(val-mean) > stdev*3:
            outlier_rows.append(counter)
        counter+=1
    outlier_rows.sort(reverse=True)
    for row in outlier_rows:
        ckd.drop(row, inplace = True)  
    ckd.reset_index(drop = True, inplace = True)
    #print(ckd)
ckd.drop('index', axis = 1, inplace = True)

# Print the dataset
print(ckd)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
1       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
2       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
3       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
4       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
128     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
129     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
130     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
131     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
132     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [9]:
# Normalize the all Numerical Attributes in the dataset.
for col in numerical_columns:
    max = ckd.loc[:, col].max()
    ckd[col] = ckd[col]/max

# Print the dataset
print(ckd)

     unique_id     age   bp       bgr        bu        sc       sod       pot  \
0       343710  0.7625  0.8  0.570957  0.907975  0.458824  0.900000  0.896552   
1       397388  0.7875  1.0  0.924092  0.214724  0.376471  0.953333  0.603448   
2       438182  0.9125  0.8  0.834983  0.871166  0.541176  0.920000  1.000000   
3       474407  0.2625  0.9  0.353135  0.245399  0.200000  0.833333  0.603448   
4       475200  0.8000  1.0  0.537954  0.331288  0.847059  0.933333  0.793103   
..         ...     ...  ...       ...       ...       ...       ...       ...   
128     486397  0.2500  0.7  0.405941  0.269939  0.117647  0.900000  0.655172   
129     203466  0.2750  0.6  0.320132  0.110429  0.141176  0.920000  0.741379   
130     482105  0.5500  0.6  0.313531  0.282209  0.058824  0.920000  0.724138   
131     228849  0.7250  0.8  0.330033  0.306748  0.141176  0.933333  0.603448   
132     996076  0.7125  0.6  0.346535  0.300613  0.141176  1.000000  0.810345   

         hemo       pcv  ..

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

In [10]:
#al and su are two columns that despite being categorical in nature, were described using numbers. 
#Because of this, it may make it more difficult to make a model or understand predictions with those columns included
#Target_notckd is the logical opposite of Target_ckd. Since we only need Target_ckd to tell us if the person had ckd or not,
#this column is unneccessary
#unique_id is not appropriate for modeling or predictions, since we are analyzing trends across the dataset, not individual patients' data. 
unnec_columns = ["al","su","Target_notckd", "unique_id"]

#Remove that column
ckd.drop(unnec_columns, axis = 1, inplace = True)

# Print the dataset
print(ckd)

        age   bp       bgr        bu        sc       sod       pot      hemo  \
0    0.7625  0.8  0.570957  0.907975  0.458824  0.900000  0.896552  0.432584   
1    0.7875  1.0  0.924092  0.214724  0.376471  0.953333  0.603448  0.730337   
2    0.9125  0.8  0.834983  0.871166  0.541176  0.920000  1.000000  0.589888   
3    0.2625  0.9  0.353135  0.245399  0.200000  0.833333  0.603448  0.466292   
4    0.8000  1.0  0.537954  0.331288  0.847059  0.933333  0.793103  0.443820   
..      ...  ...       ...       ...       ...       ...       ...       ...   
128  0.2500  0.7  0.405941  0.269939  0.117647  0.900000  0.655172  0.820225   
129  0.2750  0.6  0.320132  0.110429  0.141176  0.920000  0.741379  0.758427   
130  0.5500  0.6  0.313531  0.282209  0.058824  0.920000  0.724138  0.842697   
131  0.7250  0.8  0.330033  0.306748  0.141176  0.933333  0.603448  0.786517   
132  0.7125  0.6  0.346535  0.300613  0.141176  1.000000  0.810345  0.882022   

          pcv      wbcc  ...  dm_yes  c

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [11]:
# Export the dataframe to a new csv file
ckd.to_csv("csv_results.csv")

# Export the dataframe to a new json file
ckd.to_json("json_results.json")

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [12]:
print(ckd.head(15))

       age   bp       bgr        bu        sc       sod       pot      hemo  \
0   0.7625  0.8  0.570957  0.907975  0.458824  0.900000  0.896552  0.432584   
1   0.7875  1.0  0.924092  0.214724  0.376471  0.953333  0.603448  0.730337   
2   0.9125  0.8  0.834983  0.871166  0.541176  0.920000  1.000000  0.589888   
3   0.2625  0.9  0.353135  0.245399  0.200000  0.833333  0.603448  0.466292   
4   0.8000  1.0  0.537954  0.331288  0.847059  0.933333  0.793103  0.443820   
5   0.7500  0.6  0.950495  0.220859  0.200000  0.866667  0.517241  0.443820   
6   0.7000  0.9  0.425743  0.656442  0.788235  0.873333  0.827586  0.511236   
7   0.8625  0.7  0.706271  0.588957  0.741176  0.800000  0.672414  0.528090   
8   0.5750  0.6  0.537954  0.564417  0.388235  0.940000  0.689655  0.550562   
9   0.7750  0.7  0.402640  0.257669  0.200000  0.906667  0.810345  0.707865   
10  0.6875  0.8  0.706271  0.447853  0.458824  0.913333  0.844828  0.612360   
11  0.8000  0.6  0.788779  0.355828  0.505882  0.913

Prompt used: "Here is a dataset. Please convert it to a markdown table." [Pasted Dataset below]

### Paste the markdown table here
Note: The cleaned dataset had too many columns to display in the terminal. It was copied including the ellipsis that represents collapsed (not visible) columns.

|   age   |   bp   |     bgr     |      bu      |      sc      |     sod      |     pot      |    hemo     |    pcv     |     wbcc    | ... |  dm_yes  | cad_no | cad_yes | appet_good | appet_poor |  pe_no  | pe_yes | ane_no | ane_yes | Target_ckd |
|---------|--------|-------------|--------------|--------------|--------------|--------------|--------------|------------|-------------|-----|----------|--------|---------|------------|------------|---------|--------|--------|---------|-------------|
| 0.7625  |  0.8   | 0.570957    | 0.907975     | 0.458824     | 0.900000     | 0.896552     | 0.432584     | 0.444444   | 0.550898    | ... |   True   |  False |  True   |   False    |    True    |  False  |  True  | False  |  True   |     True    |
| 0.7875  |  1.0   | 0.924092    | 0.214724     | 0.376471     | 0.953333     | 0.603448     | 0.730337     | 0.740741   | 0.586826    | ... |  False   |  False |  True   |    True    |   False    |   True  |  False |  True  |  False  |     True    |
| 0.9125  |  0.8   | 0.834983    | 0.871166     | 0.541176     | 0.920000     | 1.000000     | 0.589888     | 0.611111   | 0.431138    | ... |   True   |  False |  True   |    True    |   False    |   True  |  False |  True  |  False  |     True    |
| 0.2625  |  0.9   | 0.353135    | 0.245399     | 0.200000     | 0.833333     | 0.603448     | 0.466292     | 0.425926   | 0.742515    | ... |  False   |   True |  False  |    True    |   False    |   True  |  False | False  |  True   |     True    |
| 0.8000  |  1.0   | 0.537954    | 0.331288     | 0.847059     | 0.933333     | 0.793103     | 0.443820     | 0.481481   | 0.449102    | ... |   True   |   True |  False  |    True    |   False    |  False  |  True  |  True  |  False  |     True    |
| 0.7500  |  0.6   | 0.950495    | 0.220859     | 0.200000     | 0.866667     | 0.517241     | 0.443820     | 0.462963   | 0.910180    | ... |  False   |   True |  False  |   False    |    True    |   True  | False  | False  |  True   |     True    |
| 0.7000  |  0.9   | 0.425743    | 0.656442     | 0.788235     | 0.873333     | 0.827586     | 0.511236     | 0.537037   | 0.383234    | ... |  False   |   True |  False  |    True    |   False    |   True  |  False |  True  |  False  |     True    |
| 0.8625  |  0.7   | 0.706271    | 0.588957     | 0.741176     | 0.800000     | 0.672414     | 0.528090     | 0.518519   | 0.688623    | ... |   True   |  False |  True   |    True    |   False    |  False  |  True  | False  |  True   |     True    |
| 0.5750  |  0.6   | 0.537954    | 0.564417     | 0.388235     | 0.940000     | 0.689655     | 0.550562     | 0.518519   | 0.874251    | ... |   True   |   True |  False  |    True    |   False    |   True  | False  |  True  |  False  |     True    |
| 0.7750  |  0.7   | 0.402640    | 0.257669     | 0.200000     | 0.906667     | 0.810345     | 0.707865     | 0.722222   | 0.473054    | ... |   True   |   True |  False  |    True    |   False    |   True  | False  |  True  |  False  |     True    |
| 0.6875  |  0.8   | 0.706271    | 0.447853     | 0.458824     | 0.913333     | 0.844828     | 0.612360     | 0.629630   | 0.443114    | ... |   True   |   True |  False  |    True    |   False    |  False  |  True  |  True  |  False  |     True    |
| 0.8000  |  0.6   | 0.788779    | 0.355828     | 0.505882     | 0.913333     | 0.931034     | 0.533708     | 0.537037   | 0.449102    | ... |   True   |   True |  False  |   False    |    True    |  False  |  True  |  True  |  False  |     True    |
| 0.8875  |  0.6   | 0.389439    | 0.766871     | 0.623529     | 0.906667     | 0.844828     | 0.640449     | 0.648148   | 0.910180    | ... |   True   |   True |  False  |   False    |    True    |  False  |  True  |  True  |  False  |     True    |
| 0.8875  |  0.7   | 0.722772    | 0.503067     | 0.423529     | 0.886667     | 0.758621     | 0.584270     | 0.611111   | 0.335329    | ... |   True   |  False |  True   |    True    |   False    |   True  | False  |  True  |  False  |     True    |
| 0.7375  |  0.8   | 1.000000    | 0.214724     | 0.152941     | 0.813333     | 0.603448     | 0.584270     | 0.648148   | 0.652695    | ... |   True   |   True |  False  |   False    |    True    |   True  | False  |  True  |  False  |     True    |

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

In LLM conversations, I could mitigate data conversion hallucinations by limiting the amount of data input at once. I could also prompt it multiple times with similar prompts and cross-check to see if there are any data irregularities.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

Prompt used: "Here is a mapping between abbreviations and meanings for columns in a dataframe named ckd. Please write a PANDAS script to rename the abbreviations to the meanings for all columns in the dataset." [abbreviation --> naming map pasted below this]

In [13]:
# Code to rename all the columns in the dataset
### Below this line, all code has been generated by 383GPT, unless otherwise noted. ----------------------------------
# Mapping of abbreviations to full meanings
rename_mapping = {
    'age': 'Age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wbcc': 'white blood cell count',
    'rbcc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    #----renamed 'class' to 'Target' as it is named differently in the dataframe than the mapping
    'Target': 'ckd or not ckd'
}

# Rename the columns using the mapping
ckd.rename(columns=rename_mapping, inplace=True)

# Display the renamed DataFrame
print(ckd)

        Age  blood pressure  blood glucose random  blood urea  \
0    0.7625             0.8              0.570957    0.907975   
1    0.7875             1.0              0.924092    0.214724   
2    0.9125             0.8              0.834983    0.871166   
3    0.2625             0.9              0.353135    0.245399   
4    0.8000             1.0              0.537954    0.331288   
..      ...             ...                   ...         ...   
128  0.2500             0.7              0.405941    0.269939   
129  0.2750             0.6              0.320132    0.110429   
130  0.5500             0.6              0.313531    0.282209   
131  0.7250             0.8              0.330033    0.306748   
132  0.7125             0.6              0.346535    0.300613   

     serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
0            0.458824  0.900000   0.896552    0.432584            0.444444   
1            0.376471  0.953333   0.603448    0.730337         

This code ran successfully, and worked for the most part. However, it failed to rename any of the categorical columns, as they had been encoded into multiple dummy columns that did not have the exact same names as the abbreviation in the abbreviation --> meaning map provided.

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

Prompt used: "Below is a SQL query. Please convert it into a pandas query." [pasted query below]

In [14]:
# Converted SQL to Pandas code
### Below this line, all code has been generated by 383GPT, unless otherwise noted. ----------------------------------
# Grouping by 'Target' and counting occurrences
#----I have renamed "df" to "ckd", as no specific dataframe name was provided in the query we were supposed to give to 383GPT
#----I have also renamed 'Target' to 'Target_ckd', as it was renamed/split so when encoding categorical data.
result = ckd.groupby('Target_ckd').size().reset_index(name='count')

print(result)

   Target_ckd  count
0       False    111
1        True     22


*Describe what the above code does here*

The code makes a new dataframe called result, which has counts based on the grouped number of rows for each value of Target_ckd (True or False) in ckd.
Result has two columns [Target_ckd, and count], and two rows. One row has the Target_ckd value of false and its associated count is how many people in the cleaned dataset did not have ckd. The second row has the Target_ckd value of True and its associated count is how many people in the cleaned dataset had ckd.